jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java

```diff
@@ -121,85 +124,106 @@ public boolean clear()
-                {
-                    if (!cookieDomain.equals("localhost"))
-                        return false;
-                }
-
-                String domain = uri.getHost();
-                if (domain != null)
-                {
-                    domain = domain.toLowerCase(Locale.ENGLISH);
-                    // If uri.host==foo.example.com, only accept
-                    // cookie.domain==(foo.example.com|example.com).
-                    if (!domain.endsWith(cookieDomain))
-                        return false;
-                    int beforeMatch = domain.length() - cookieDomain.length() - 1;
-                    if (beforeMatch >= 0 && domain.charAt(beforeMatch) != '.')
-                        return false;
-                }
-            }
-            else
-            {
-                // No explicit cookie domain, use the origin domain.
-                cookieDomain = uri.getHost();
-            }
+            String resolvedPath = resolvePath(uri, cookie);
 
-            // Cookies are stored under their domain, so that:
-            // - add(sub.example.com, cookie[Domain]=null) => Key[domain=sub.example.com]
-            // - add(sub.example.com, cookie[Domain]=example.com) => Key[domain=example.com]
+            // Cookies are stored under their resolved domain, so that:
+            // - add(sub.example.com, cookie[Domain]=null) => key=sub.example.com
+            // - add(sub.example.com, cookie[Domain]=example.com) => key=example.com
             // This facilitates the matching algorithm.
-            Key key = new Key(secure, cookieDomain);
-            boolean[] result = new boolean[]{true};
+            boolean[] added = new boolean[1];
+            StoredHttpCookie storedCookie = new StoredHttpCookie(cookie, uri, resolvedDomain, resolvedPath);
             try (AutoLock ignored = lock.lock())
             {
+                String key = resolvedDomain.toLowerCase(Locale.ENGLISH);
                 cookies.compute(key, (k, v) ->
                 {
                     // RFC 6265, section 4.1.2.
                     // Evict an existing cookie with
                     // same name, domain and path.
                     if (v != null)
-                        v.remove(cookie);
+                        v.remove(storedCookie);
 
                     // Add only non-expired cookies.
                     if (cookie.isExpired())
-                    {
-                        result[0] = false;
                         return v == null || v.isEmpty() ? null : v;
-                    }
 
+                    added[0] = true;
                     if (v == null)
                         v = new ArrayList<>();
-                    v.add(new Cookie(cookie));
+                    v.add(storedCookie);
                     return v;
                 });
             }
 
-            return result[0];
+            return added[0];
+        }
+
+        private String resolveDomain(URI uri, HttpCookie cookie)
```

Comment:
Some test cases to consider ...

An http uri without an authority.
I think we should reject this.

```
jshell> var foo = URI.create("http:///path")
foo ==> http:///path

jshell> foo.getHost()
$2 ==> null

jshell> foo.isOpaque()
$3 ==> false

jshell> foo.isAbsolute()
$4 ==> true

jshell> foo.getAuthority()
$5 ==> null
```

A URI with an authority that has no host.
I think we should reject this as well.

```
jshell> var foo = URI.create("http://user@/path")
foo ==> http://user@/path

jshell> foo.getHost()
$2 ==> null

jshell> foo.getAuthority()
$3 ==> "user@"

jshell> foo.isAbsolute()
$4 ==> true
```

A non-absolute URI
I think we should reject this too.

```
jshell> var foo = URI.create("/path")
foo ==> /path

jshell> foo.getHost()
$2 ==> null

jshell> foo.isAbsolute()
$3 ==> false

jshell> foo.getPath()
$4 ==> "/path"
```

Reply:
Already covered, all the examples have a `null` URI host.